### PR TITLE
feat: support actuator for sct core components

### PR DESCRIPTION
**Pull request by shuiqingliu**
This commit only support actuator of config, metadata, discovery and ratelimit.

## PR Type
Feature.

## Describe what this PR does for and how you did.
This PR has extend actuator of Spring Cloud Tencent. It only contains base info for core module.

## Adding the issue link (#xxx) if possible.
#243 

## Checklist

- [x] Add copyright holder at the beginning of .class file if it is new.
- [ ] Add information of this PR to CHANGELOG.md in root of project.
- [x] All junit tests passing.
- [x] Coverage from `Codecov Report` should not decrease.

## Checklist (Optional)

- [x] Will Pull Request to branch of 2020.0 and 2021.0.
- [x] Add documentation in javadoc in code or comment below the PR if necessary.
- [x] Add your name as @author to the beginning of .class file.

### CODE REVIEW

**Comment by misselvexu** [comment]
@shuiqingliu  

Thank you for your first contribution PR ，Can provide a sample of the final output of all Actuator Endpoints ~

**Comment by shuiqingliu** [comment]
@misselvexu I extract the final output from example module as follows.
PS: `ratelimit endpoint` has no fully test. I have not found a way to do that.

## 1. Service Discovery Endpoint
* url: ip:port/actuator/polaris-discovery

// http://127.0.0.1:48080/actuator/polaris-discovery/DiscoveryCalleeService
```json
{
  "PolarisDiscoveryProperties": {
    "namespace": "default",
    "service": "DiscoveryCallerService",
    "token": null,
    "weight": 100,
    "version": null,
    "protocol": "http",
    "port": 48080,
    "enabled": true,
    "registerEnabled": true,
    "heartbeatEnabled": true,
    "healthCheckUrl": "/discovery/service/caller/healthCheck",
    "serviceListRefreshInterval": 60000
  },
  "ServiceInstance": [
    {
      "scheme": "http",
      "uri": "http://10.97.61.101:48084",
      "secure": false,
      "instanceId": null,
      "metadata": {
        "k111111": "v1",
        "region": "application",
        "zone": "zone-z",
        "version": "",
        "protocol": "http",
        "boot": "true",
        "internal-address": "10.97.61.101:48084",
        "internal-ip": "10.97.61.101"
      },
      "serviceId": "DiscoveryCalleeService",
      "port": 48084,
      "host": "10.97.61.101"
    },
    {
      "scheme": "http",
      "uri": "http://192.168.1.72:48084",
      "secure": false,
      "instanceId": null,
      "metadata": {
        "internal-address": "192.168.1.72:48084",
        "internal-ip": "192.168.1.72",
        "k1": "v1",
        "zone": "shanghai",
        "version": "",
        "protocol": "http"
      },
      "serviceId": "DiscoveryCalleeService",
      "port": 48084,
      "host": "192.168.1.72"
    }
  ]
}
```
## 2. Polaris Ratelimit Endpoint

// http://127.0.0.1:48080/actuator/ratelimit/{namespace}/{serviceId}?dstService="test"

```json
{
  "rateLimits": [
    
  ],
  "service": "test",
  "namespace": "default",
  "properties": {
    "rejectRequestTips": null,
    "rejectRequestTipsFilePath": "reject-tips.html",
    "rejectHttpCode": 429,
    "maxQueuingTime": 500
  }
}
````

## 3. Polaris Config Endpoint

// http://127.0.0.1:48080/actuator/polaris-config
```json
{
  "PolarisPropertySource": [
    {
      "name": "default-polaris-config-example-config/application.properties",
      "source": {
        
      },
      "namespace": "default",
      "group": "polaris-config-example",
      "fileName": "config/application.properties",
      "propertySourceName": "default-polaris-config-example-config/application.properties",
      "propertyNames": [
        
      ]
    },
    {
      "name": "default-polaris-config-example-config/bootstrap.yml",
      "source": {
        
      },
      "namespace": "default",
      "group": "polaris-config-example",
      "fileName": "config/bootstrap.yml",
      "propertySourceName": "default-polaris-config-example-config/bootstrap.yml",
      "propertyNames": [
        
      ]
    },
    {
      "name": "default-polaris-config-example-bootstrap.properties",
      "source": {
        
      },
      "namespace": "default",
      "group": "polaris-config-example",
      "fileName": "bootstrap.properties",
      "propertySourceName": "default-polaris-config-example-bootstrap.properties",
      "propertyNames": [
        
      ]
    },
    {
      "name": "default-polaris-config-example-bootstrap.yml",
      "source": {
        
      },
      "namespace": "default",
      "group": "polaris-config-example",
      "fileName": "bootstrap.yml",
      "propertySourceName": "default-polaris-config-example-bootstrap.yml",
      "propertyNames": [
        
      ]
    },
    {
      "name": "default-polaris-config-example-application.yml",
      "source": {
        "teacher.age": "25",
        "teacher.name": "Skye"
      },
      "namespace": "default",
      "group": "polaris-config-example",
      "fileName": "application.yml",
      "propertySourceName": "default-polaris-config-example-application.yml",
      "propertyNames": [
        "teacher.age",
        "teacher.name"
      ]
    },
    {
      "name": "default-polaris-config-example-application.properties",
      "source": {
        "timeout": "888"
      },
      "namespace": "default",
      "group": "polaris-config-example",
      "fileName": "application.properties",
      "propertySourceName": "default-polaris-config-example-application.properties",
      "propertyNames": [
        "timeout"
      ]
    }
  ],
  "PolarisConfigProperties": {
    "enabled": true,
    "address": null,
    "port": 8093,
    "autoRefresh": true,
    "groups": [
      {
        "name": "polaris-config-example",
        "files": [
          "config/application.properties",
          "config/bootstrap.yml"
        ]
      }
    ]
  }
}
```

## 4. Metadata Endpoint
// http://127.0.0.1:48080/actuator/metadata

```json
{
  "zone": null,
  "Config": {
    "CUSTOM-METADATA-KEY-TRANSITIVE": "CUSTOM-VALUE-TRANSITIVE",
    "CUSTOM-METADATA-KEY-LOCAL": "CUSTOM-VALUE-LOCAL"
  },
  "campus": null,
  "ConfigTransitive": {
    "CUSTOM-METADATA-KEY-TRANSITIVE": "CUSTOM-VALUE-TRANSITIVE"
  },
  "MergeStatic": {
    "CUSTOM-METADATA-KEY-TRANSITIVE": "CUSTOM-VALUE-TRANSITIVE",
    "CUSTOM-METADATA-KEY-LOCAL": "CUSTOM-VALUE-LOCAL"
  },
  "Env": {
    
  },
  "region": null,
  "EnvTransitive": {
    
  }
}
```

**Comment by misselvexu** [comment]
@shuiqingliu  Please resolve conflicts .

**Comment by shuiqingliu** [comment]
@misselvexu done.

File: spring-cloud-starter-tencent-polaris-ratelimit/src/main/java/com/tencent/cloud/polaris/ratelimit/endpoint/PolarisRateLimitRuleEndpointAutoConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import org.springframework.context.annotation.Configuration;
+
+/**
+ * The AutoConfiguration for Polaris RateLimit endpoint.
+ *
+ * @author shuiqingliu
+ **/
+
+@Configuration(proxyBeanMethods = false)
+@ConditionalOnClass(Endpoint.class)
```
Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-ratelimit/src/main/java/com/tencent/cloud/polaris/ratelimit/endpoint/PolarisRateLimitRuleEndpointAutoConfiguration.java:39]: 建议换成类名，否则当没有引入actuator依赖时找不到对应类。

File: spring-cloud-starter-tencent-polaris-ratelimit/src/test/java/com/tencent/cloud/polaris/ratelimit/endpoint/PolarisRateLimitRuleEndpointTests.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+import static com.tencent.polaris.test.common.Consts.PORT;
+import static com.tencent.polaris.test.common.Consts.SERVICE_PROVIDER;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+/**
+ *	Test for polaris rete limit rule endpoint.
```
Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-ratelimit/src/test/java/com/tencent/cloud/polaris/ratelimit/endpoint/PolarisRateLimitRuleEndpointTests.java:47]: 删除多余的空格

File: spring-cloud-starter-tencent-polaris-ratelimit/src/test/java/com/tencent/cloud/polaris/ratelimit/endpoint/PolarisRateLimitRuleEndpointTests.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+						.setValueType(ModelProto.MatchString.ValueType.TEXT).build();
+				RateLimitProto.Rule rule = RateLimitProto.Rule.newBuilder()
+						.putLabels("${http.method}", matchString).build();
+				return RateLimitProto.RateLimit.newBuilder().addRules(rule).build();
+			}
+		});
+	}
+
+	@Test
+	public void polarisRateLimit() {
```
Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-ratelimit/src/test/java/com/tencent/cloud/polaris/ratelimit/endpoint/PolarisRateLimitRuleEndpointTests.java:99]: 改成`testPolarisRateLimit`

File: spring-cloud-starter-tencent-polaris-config/src/test/java/com/tencent/cloud/polaris/config/endpoint/PolarisConfigEndpointTests.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+	private final String testFileName = "application.properties";
+
+
+	@Mock
+	private PolarisConfigProperties polarisConfigProperties;
+	@Mock
+	private PolarisPropertySourceManager polarisPropertySourceManager;
+
+	@Test
+	public void polarisConfigEndpoint() {
```
Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-config/src/test/java/com/tencent/cloud/polaris/config/endpoint/PolarisConfigEndpointTests.java:58]: 方法建议添加`test`前缀

File: spring-cloud-starter-tencent-polaris-discovery/src/test/java/com/tencent/cloud/polaris/endpoint/PolarisDiscoveryEndPointTests.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+	@AfterClass
+	public static void afterClass() throws Exception {
+		if (null != namingServer) {
+			namingServer.terminate();
+		}
+	}
+
+
+	@Test
+	public void polarisDiscoveryEndpoint() {
```
Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-discovery/src/test/java/com/tencent/cloud/polaris/endpoint/PolarisDiscoveryEndPointTests.java:80]: 方法建议添加`test`前缀

File: spring-cloud-tencent-commons/src/main/java/com/tencent/cloud/common/metadata/endpoint/MetadataEndpoint.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+	private final StaticMetadataManager staticMetadataManager;
+
+	public MetadataEndpoint(StaticMetadataManager staticMetadataManager) {
+		this.staticMetadataManager = staticMetadataManager;
+	}
+
+	@ReadOperation
+	public Map<String, Object> metadate() {
+		Map<String, Object>  result = new HashMap<>();
+		result.put("Env", staticMetadataManager.getAllEnvMetadata());
```
Comment (SkyeBeFreeman) [inline spring-cloud-tencent-commons/src/main/java/com/tencent/cloud/common/metadata/endpoint/MetadataEndpoint.java:46]: 可以拉取新的代码，新的支持开发者使用接口的方式添加元数据——`customSPIMetadata`

File: spring-cloud-tencent-commons/src/main/java/com/tencent/cloud/common/metadata/endpoint/MetadataEndpointAutoConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+import org.springframework.context.annotation.Configuration;
+
+/**
+ * The AutoConfiguration for metadata endpoint.
+ *
+ * @author shuiqingliu
+ **/
+
+@Configuration(proxyBeanMethods = false)
+@ConditionalOnClass(Endpoint.class)
```
Comment (SkyeBeFreeman) [inline spring-cloud-tencent-commons/src/main/java/com/tencent/cloud/common/metadata/endpoint/MetadataEndpointAutoConfiguration.java:37]: 建议替换为类名

File: spring-cloud-tencent-commons/src/test/java/com/tencent/cloud/common/metadata/endpoint/MetadataEndpointTests.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+ *  @author shuiqingliu
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class MetadataEndpointTests {
+
+	@Mock
+	private StaticMetadataManager staticMetadataManager;
+
+	@Test
+	public void metadataEndpoint() {
```
Comment (SkyeBeFreeman) [inline spring-cloud-tencent-commons/src/test/java/com/tencent/cloud/common/metadata/endpoint/MetadataEndpointTests.java:46]: 建议添加`test`前缀

File: spring-cloud-starter-tencent-polaris-ratelimit/src/main/java/com/tencent/cloud/polaris/ratelimit/endpoint/PolarisRateLimitRuleEndpoint.java [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+import org.springframework.lang.Nullable;
+import org.springframework.util.CollectionUtils;
+
+/**
+ *	Endpoint of Polaris RateLimit rule.
+ *
+ * @author shuiqingliu
+ **/
+
+@Endpoint(id = "ratelimit")
```
Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-ratelimit/src/main/java/com/tencent/cloud/polaris/ratelimit/endpoint/PolarisRateLimitRuleEndpoint.java:43]: `polaris-ratelimit`是不是更好？区分出是北极星的限流

File: spring-cloud-tencent-examples/polaris-ratelimit-example/ratelimit-callee-service/src/main/resources/bootstrap.yml [hunk 10]
```diff
@@ -12,3 +12,10 @@ spring:
         enabled: true
         rejectRequestTipsFilePath: reject-tips.html
         maxQueuingTime: 500
+
+management:
+  endpoints:
+    web:
+      exposure:
+        include:
+          - ratelimit
```
Comment (SkyeBeFreeman) [inline spring-cloud-tencent-examples/polaris-ratelimit-example/ratelimit-callee-service/src/main/resources/bootstrap.yml:21]: `polaris-ratelimit`是不是更好？

File: spring-cloud-tencent-examples/metadata-transfer-example/metadata-caller-service/src/main/resources/bootstrap.yml [hunk 11]
```diff
@@ -24,3 +24,9 @@ spring:
         # Assigned which metadata key-value will be passed along the link
         transitive:
           - CUSTOM-METADATA-KEY-TRANSITIVE
+management:
+  endpoints:
+    web:
+      exposure:
+        include:
+          - metadata
```
Comment (SkyeBeFreeman) [inline spring-cloud-tencent-examples/metadata-transfer-example/metadata-caller-service/src/main/resources/bootstrap.yml:32]: 这里会不会引起歧义？因为spring cloud自带也有metadata

**Review by SkyeBeFreeman — CHANGES_REQUESTED** [review]

**Review by misselvexu — CHANGES_REQUESTED** [review]
@shuiqingliu 
1. Check the definition of all the Endpoint URL, unify the naming rules.
2. Check all requests JSON response value of the JSON-Key case specification unified, eg: hump, underscore, initial capitalization, etc.

File: spring-cloud-tencent-commons/src/main/java/com/tencent/cloud/common/metadata/endpoint/PolarisMetadataEndpointAutoConfiguration.java [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+/**
+ * The AutoConfiguration for metadata endpoint.
+ *
+ * @author shuiqingliu
+ **/
+@Configuration(proxyBeanMethods = false)
+@ConditionalOnClass(PolarisMetadataEndpoint.class)
```
Comment (SkyeBeFreeman) [inline spring-cloud-tencent-commons/src/main/java/com/tencent/cloud/common/metadata/endpoint/PolarisMetadataEndpointAutoConfiguration.java:34]: 改成`@ConditionalOnClass(Endpoint.class)`

File: spring-cloud-starter-tencent-polaris-config/src/main/java/com/tencent/cloud/polaris/config/endpoint/PolarisConfigEndpointAutoConfiguration.java [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+/**
+ * The AutoConfiguration for Polaris Config's endpoint.
+ *
+ * @author shuiqingliu
+ **/
+@Configuration(proxyBeanMethods = false)
+@ConditionalOnClass(PolarisConfigEndpoint.class)
```
Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-config/src/main/java/com/tencent/cloud/polaris/config/endpoint/PolarisConfigEndpointAutoConfiguration.java:37]: 改成`@ConditionalOnClass(Endpoint.class)`

**Review by SkyeBeFreeman — CHANGES_REQUESTED** [review]

**Review by SkyeBeFreeman — APPROVED** [review]
LGTM.

**Comment by misselvexu** [comment]
@shuiqingliu  
Will you cherry-pick to branch 2020.0 and 2021.0 ？

**Comment by shuiqingliu** [comment]
> 

yes. I will

**Pull request merged**